productinfo/ec2/productinfo_ec2.go

```diff
@@ -159,6 +175,102 @@ func (e *Ec2Infoer) GetProducts(regionId string, attrKey string, attrValue produ
 	return vms, nil
 }
 
+type priceData struct {
+	awsData aws.JSONValue
+	attrMap map[string]interface{}
+}
+
+func newPriceData(prData aws.JSONValue) (*priceData, error) {
+	pd := priceData{awsData: prData}
+
+	// get the attributes map
+	productMap, err := getMapForKey("product", pd.awsData)
+	if err != nil {
+		return nil, err
+	}
+
+	attrMap, err := getMapForKey("attributes", productMap)
+	if err != nil {
+		return nil, err
+	}
+
+	pd.attrMap = attrMap
+
+	return &pd, nil
+}
+func (pd *priceData) GetDataForKey(attr string) (string, error) {
+	switch attr {
+	case "instanceType":
+		instanceType, ok := pd.attrMap["instanceType"].(string)
+		if !ok {
+			return "", errors.New("could not get instance type or could not cast instance type to string")
+		}
+		return instanceType, nil
+	case Cpu:
+		vcpu, ok := pd.attrMap[Cpu].(string)
+		if !ok {
+			return "", errors.New("could not get vcpu or could not cast vcpu to string")
+		}
+		return vcpu, nil
+	case Memory:
+		mem, ok := pd.attrMap[Memory].(string)
+		if !ok {
+			return "", errors.New("could not get memory or could not cast memory to string")
+		}
+		return mem, nil
+	case "gpu":
+		gpu, ok := pd.attrMap["gpu"].(string)
+		if !ok {
+			return "", errors.New("could not get gpu or could not cast gpu to string")
+		}
+		return gpu, nil
+	}
+	return "", errors.New("invalid attribute")
+}
```

Comment:
this function is still too complicated, it can be done very simply (invalid attributes are not checked but I don't think that we really need that):
```
func (pd *priceData) GetDataForKey(attr string) (string, error) {
	if value, ok := pd.attrMap[attr].(string); !ok {
		return "", fmt.Errorf("could not get %s or could not cast %s to string", attr, attr)
	} else {
		return value, nil
	}
}
```

Reply:
also please squash these commits before merge

Reply:
+1, 
I'd rather write these in the following form:
```func (pd *priceData) GetDataForKey(attr string) (string, error) {
	if value, ok := pd.attrMap[attr].(string); ok {
              return value, nil
	} 
        return "", fmt.Errorf("could not get %s or could not cast %s to string", attr, attr)
}
```